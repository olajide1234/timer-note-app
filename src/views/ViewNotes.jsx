import React from 'react'
import { ViewSingleNote } from '../components'
import { useSelector, useDispatch } from "react-redux";

/**
 * Container to fetch and display all notes
 *
 * @component
 * @example
 * return (
 *   <ViewNotes />
 * )
 */
export function ViewNotes() {
  const dispatch = useDispatch();
  const notes = useSelector(state => state);

  /**
   * Function to delete a note 
   *
   * @param   {int} id Id of note to be deleted
   * @returns  {void}
   */
  const deleteFn = (id) => dispatch({ type: 'DELETE_NOTE', deleteId: id })

  return (
    <div>
      <div className='headerText'>View and Delete Notes</div>
      <div className='noteContainer'>
        {notes.length < 1 ? 'Add notes to see them appear here' : [...notes].reverse().map((note, i) => <ViewSingleNote key={i} id={note.id} time={note.time} text={note.text} deleteFn={deleteFn} />)}
      </div>
    </div>
  );
}
