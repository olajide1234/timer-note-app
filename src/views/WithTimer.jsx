import React, { Fragment, useState } from 'react'
import { AddNote, Timer } from '../components'
import { Col } from 'antd';

var hours = 0;
var mins = 0;
var seconds = 0;

/**
 * Function to start timer
 * Because JS does not pass-by-reference and we need to persist changes to hrs, mins and secs, we don't pass these as arguments.
 * We also simulate overloading on this function for testing
 * @param   {function} updateTimerHrs Function to update hour of timer
 * @param   {function} updateTimerMins Function to update minutes of timer
 * @param   {function} updateTimerSecs Function to update seconds of timer
 * @param   {function} updateTimerStopId Function to id of timer for stopping
 * @param   {int} h Current hour for overloading/testing
 * @param   {int} m Current minute for overloading/testing
 * @param   {int} s Current secoond for overloading/testing
 * @returns  {void} 
 */

export const startTimer = (updateTimerHrs, updateTimerMins, updateTimerSecs, updateTimerStopId, h, m, s) => {
  var stopId = setTimeout(() => {
    s ? s++ : seconds++;

    if (seconds > 59 || s > 59) {
      s ? s = 0 : seconds = 0; m ? m++ : mins++;
      if (mins > 59 || m > 59) {
        m ? m = 0 : mins = 0; h ? h++ : hours++;
        if (h) {
          if (h < 10) { updateTimerHrs(`0${hours}:`); } else { updateTimerHrs(`${hours}:`); }
        } else {
          if (hours < 10) { updateTimerHrs(`0${hours}:`); } else { updateTimerHrs(`${hours}:`); }
        }
      }

      if (m) {
        if (m < 10) {
          updateTimerMins(`0${mins}:`);
        }
        else {
          updateTimerMins(`${mins}:`);
        }
      } else {
        if (mins < 10) {
          updateTimerMins(`0${mins}:`);
        }
        else {
          updateTimerMins(`${mins}:`);
        }
      }

    }
    if (seconds < 10 || s < 10) {
      updateTimerSecs(`0${seconds}`);
    } else {
      updateTimerSecs(`${seconds}`);
    }
    updateTimerStopId(stopId)
    startTimer(updateTimerHrs, updateTimerMins, updateTimerSecs, updateTimerStopId);
  }, 1000);
}

/**
 * Resets timer
 *
 * @param   {int} id ID of current set timeout
 * @param   {function} updateTimerHrs Function to update hour of timer
 * @param   {function} updateTimerMins Function to update minutes of timer
 * @param   {function} updateTimerSecs Function to update seconds of timer
 * @returns  {void}
 */
export const resetTimer = (id, updateTimerHrs, updateTimerMins, updateTimerSecs) => {
  clearTimeout(id);
  hours = 0; mins = 0; seconds = 0;
  updateTimerSecs('00');
  updateTimerMins('00:');
  updateTimerHrs('00:');
}

/**
 * Container to generate timer and pass to child components
 *
 * @component
 * @example
 * return (
 *   <WithTimer />
 * )
 */
export function WithTimer() {

  const [timerStopId, updateTimerStopId] = useState(0);
  const [timerSecs, updateTimerSecs] = useState('00');
  const [timerMins, updateTimerMins] = useState('00:');
  const [timerHrs, updateTimerHrs] = useState('00:');
  const [timerPause, setTimerPause] = useState(false);

  /**
   * Calls function to start timer
   *
   * @returns  {void} 
   */
  const startFn = () => { startTimer(updateTimerHrs, updateTimerMins, updateTimerSecs, updateTimerStopId); setTimerPause(false) }

  /**
   * Function to stop timer
   *
   * @param   {int} id ID of current set timeout
   * @returns  {void}
   */
  const stopFn = (id) => { clearTimeout(id); setTimerPause(true) };

  /**
   * Calls function to reset timer
   *
   * @param   {int} id ID of current set timeout
   * @returns  {void}
   */
  const resetFn = (id) => { setTimerPause(false); resetTimer(id, updateTimerHrs, updateTimerMins, updateTimerSecs); }

  return (
    <Fragment>
      <Col>
        <Timer
          time={`${timerHrs}${timerMins}${timerSecs}`}
          startFn={startFn}
          stopFn={stopFn}
          resetFn={resetFn}
          currentTimerStopId={timerStopId + 1}
          timerPause={timerPause}
        />
      </Col>
      <Col><AddNote time={`${timerHrs}${timerMins}${timerSecs}`} stopFn={stopFn} currentTimerStopId={timerStopId + 1} /></Col>
    </Fragment>
  );
}
