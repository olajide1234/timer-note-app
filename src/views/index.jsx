export { ViewNotes } from './ViewNotes';
export { WithTimer } from './WithTimer'
