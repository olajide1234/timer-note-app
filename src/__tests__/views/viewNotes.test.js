import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store'
import { ViewNotes } from '../../views';
import renderer from 'react-test-renderer';

const mockStore = configureMockStore();

it('should view single note with sample', () => {
  const store = mockStore([{ id: 1, time: '12:12:12', text: 'Here is a sample note. Add more notes to see your note above this.' }]);
  const tree = renderer.create(
    <Provider store={store}>
      <ViewNotes time='11:11:11' />
    </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should view single note without sample', () => {
  const store = mockStore([]);
  const tree = renderer.create(
    <Provider store={store}>
      <ViewNotes time='11:11:11' />
    </Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});
