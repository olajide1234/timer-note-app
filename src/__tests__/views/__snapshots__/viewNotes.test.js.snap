// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should view single note with sample 1`] = `
<div>
  <div
    className="headerText"
  >
    View and Delete Notes
  </div>
  <div
    className="noteContainer"
  >
    <div
      className="smallerTopMargin"
    >
      <div
        className="ant-card ant-card-bordered ant-card-small"
        style={
          Object {
            "width": 250,
          }
        }
      >
        <div
          className="ant-card-head"
          style={Object {}}
        >
          <div
            className="ant-card-head-wrapper"
          >
            <div
              className="ant-card-head-title"
            >
              12:12:12
            </div>
            <div
              className="ant-card-extra"
            >
              <button
                className="ant-btn ant-btn-danger ant-btn-sm"
                onClick={[Function]}
                type="button"
              >
                <span>
                  Delete
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <p>
            Here is a sample note. Add more notes to see your note above this.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should view single note without sample 1`] = `
<div>
  <div
    className="headerText"
  >
    View and Delete Notes
  </div>
  <div
    className="noteContainer"
  >
    Add notes to see them appear here
  </div>
</div>
`;
